subsys/shell/shell.c

```diff
@@ -8,7 +8,9 @@
 #include <stdlib.h>
 #include <sys/atomic.h>
 #include <shell/shell.h>
+#if defined(CONFIG_SHELL_BACKEND_DUMMY)
```

Comment:
why this change is needed? API's shall not be ifdefed (https://docs.zephyrproject.org/latest/guides/coding_guidelines/index.html#rule-a-1-conditional-compilation) so ifdefing headers should not be needed as well.

Reply:
This does not explicitly state so:`
>
>Do not conditionally compile function declarations _in header files_. Do not conditionally compile structure declarations _in header files_. You may conditionally exclude fields within structure definitions to avoid wasting memory when the feature they support is not enabled

It is nowhere there that the headers shall not be excluded, as far as I can tell  and all sentences are about ifdefs **within** header files.

There might be  implicit suggestion in rationale:
>
>Excluding declarations _from the header_ based on compile-time options may prevent their documentation from being generated. 

but it again says: **from the header**.

I can fix that by defining intermediate variable when `CONFIG_SHELL_BACKEND_DUMMY_BUF_SIZE` is not defined because the dummy backend has not been selected, if you prefer that way.

There is precedence in code where the headers are ifdefed, for example in `./bluetooth/controller/hci/hci_driver.c`
```
...
#include <drivers/bluetooth/hci_driver.h>

#ifdef CONFIG_CLOCK_CONTROL_NRF
#include <drivers/clock_control/nrf_clock_control.h>
#endif
...
```

or `./subsys/net/lib/lwm2m/lwm2m_engine.c`
```
...
#include <net/socket.h>
#if defined(CONFIG_LWM2M_DTLS_SUPPORT)
#include <net/tls_credentials.h>
#endif
#if defined(CONFIG_DNS_RESOLVER)
#include <net/dns_resolve.h>
#endif
...
```

Reply:
ah, ok. I see that buf size is not defined there. I will not block but personally i prefer to create ifdef.